### ISSUE TITLE
Emit compile-time error for unsafe method calls

### ISSUE THREAD

**Issue by achugaev93** [Microsoft/TypeScript#23505]
Compiler should detect calls of **unsafe methods** (methods that can throw errors) and emit compile-time error. This feature will protect against **runtime errors** because currenly it worth nothing to miss unsafe calls.

Unsafe (known) methods are those ones who
- marked with _@throws <ErrorType>_ annotation in JSDoc
- contains `throw` operator inside method's body

This functionality is some kind of Java's `throws` clause (https://beginnersbook.com/2013/04/java-throws/).

Also code assistance (code generation) would be very helpful:
- Automatically wrap code with `try-catch` block
- Add/update JSDoc comment and add @throws annotation there

**Code**

```ts
export class FileSystem implements FileStorage {
    /**
     * @throws FileNotFoundException if such file does not exist
     */
    public async readAllText(filePath: string): Promise<string> {
        // ...
    }
}

export class SettingsLoader {
    private readonly _storage: FileStorage;

    public constructor(storage: FileStorage) {
        this._storage = storage;
    }

    public async load(settingsFilePath: string): Promise<Map<string, string>> {
        const contents = await this._storage.readAllText(settingsFilePath); // Error: Possible unhandled FileNotFoundException. Wrap call with try-catch or specify @throws annotation in JSDoc comment for method.
        // ...
    }
```

**Expected behavior:**

Compile-time error should be emitted telling about "Possible unhandled <error type>. Wrap call with try-catch or specify @throws annotation in JSDoc comment for method.".

Code-assistance for automatic code generation will be very helpful.

**Actual behavior:**

No compile-time checks for unsafe method calls.

**Comment by andy-ms** [comment]
Duplicate of #13219

**Comment by achugaev93** [comment]
@andy-ms, it's not the same. 
In #13219 new syntax request. 
This suggestion completely based on existing syntax. No new syntax constructions needed to do this improvement.

**Comment by mhegazy** [comment]
> No new syntax constructions needed to do this improvement,

what if this method is coming form a .d.ts file, how can you convey the information about it throwing an error when there is no body for the compiler to infer from? what if the compiler is generating that .d.ts file, how can it record the information it just inferred?

A new syntax is required, just like return type annotation.

**Comment by achugaev93** [comment]
> what if the compiler is generating that .d.ts file, how can it record the information it just inferred?

Maybe compiler can generate JSDoc in .d.ts with @throws annotations?

**Comment by mhegazy** [comment]
> @mhegazy, maybe compiler can generate JSDoc in .d.ts with @throws annotations? Or even copy JSDoc from source file?

the whole idea of adding optional type syntax to JS is to avoid putting types in comments.. so i would say that is not a good solution.
